src/main/java/org/wildfly/security/auth/client/MatchPortRule.java

```diff
@@ -43,6 +43,10 @@ MatchRule reparent(final MatchRule newParent) {
     }
 
     boolean halfEqual(final MatchRule other) {
+        // If the parent is 'ALL' then this test will always fail since ALL always returns 0, making this call fail
+        if (other.getMatchPort() == 0) {
+            return other.toString().equals(this.toString()) && parentHalfEqual(other);
+        }
         return other.getMatchPort() == port && parentHalfEqual(other);
```

Comment:
Is not missing `getMatchPort()` the actual problem here?
I dont think it is good idea to compare MatchRules by toString() output.